src/simulation/dynamics/spinningBodies/spinningBodiesTwoDOF/spinningBodyTwoDOFStateEffector.cpp

```diff
@@ -0,0 +1,484 @@
+ ISC License
+
+ Copyright (c) 2022, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
+
+ Permission to use, copy, modify, and/or distribute this software for any
+ purpose with or without fee is hereby granted, provided that the above
+ copyright notice and this permission notice appear in all copies.
+
+ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+ */
+
+#include "spinningBodyTwoDOFStateEffector.h"
+#include "architecture/utilities/avsEigenSupport.h"
+#include "architecture/utilities/rigidBodyKinematics.h"
+#include <string>
+
+/*! This is the constructor, setting variables to default values */
+SpinningBodyTwoDOFStateEffector::SpinningBodyTwoDOFStateEffector()
+{
+    // Zero the mass props and mass prop rates contributions
+    this->effProps.mEff = 0.0;
+    this->effProps.rEff_CB_B.fill(0.0);
+    this->effProps.IEffPntB_B.fill(0.0);
+    this->effProps.rEffPrime_CB_B.fill(0.0);
+    this->effProps.IEffPrimePntB_B.fill(0.0);
+
+    // Initialize variables to working values
+    this->IS1PntSc1_S1.setIdentity();
+    this->IS2PntSc2_S2.setIdentity();
+    this->dcm_S10B.setIdentity();
+    this->dcm_S20S1.setIdentity();
+    this->r_S1B_B.setZero();
+    this->r_S2S1_S1.setZero();
+    this->s1Hat_S1.setZero();
+    this->s2Hat_S2.setZero();
+
+    // Create the output vectors
+    this->spinningBodyConfigLogOutMsgs.push_back(new Message<SCStatesMsgPayload>);
+    this->spinningBodyConfigLogOutMsgs.push_back(new Message<SCStatesMsgPayload>);
+    this->spinningBodyOutMsgs.push_back(new Message<HingedRigidBodyMsgPayload>);
+    this->spinningBodyOutMsgs.push_back(new Message<HingedRigidBodyMsgPayload>);
+    
+    this->nameOfTheta1State = "spinningBodyTheta1" + std::to_string(SpinningBodyTwoDOFStateEffector::effectorID);
+    this->nameOfTheta1DotState = "spinningBodyTheta1Dot" + std::to_string(SpinningBodyTwoDOFStateEffector::effectorID);
+    this->nameOfTheta2State = "spinningBodyTheta2" + std::to_string(SpinningBodyTwoDOFStateEffector::effectorID);
+    this->nameOfTheta2DotState = "spinningBodyTheta2Dot" + std::to_string(SpinningBodyTwoDOFStateEffector::effectorID);
+    SpinningBodyTwoDOFStateEffector::effectorID++;
+}
+
+uint64_t SpinningBodyTwoDOFStateEffector::effectorID = 1;
+
+/*! This is the destructor, nothing to report here */
+SpinningBodyTwoDOFStateEffector::~SpinningBodyTwoDOFStateEffector()
+{
+    // Free memory to avoid errors
+    for (auto & spinningBodyConfigLogOutMsg : this->spinningBodyConfigLogOutMsgs) {
+        free(spinningBodyConfigLogOutMsg);
```

Comment:
is this required with standard vectors?  I thought these vectors took care of their own freeing up of elements?  Or is this because you used the `new` command in the constructor?  I'm happy to follow Pat's guidance on this.

Reply:
I believe this is required because I pre-populate the vector with the `new` command in the constructor. @patkenneally can you confirm this?

Reply:
With `new` we want to use `delete`. https://isocpp.org/wiki/faq/freestore-mgmt#mixing-malloc-and-delete

That being said, this free can be removed based upon my above comment.